### PR TITLE
[pickers] Migrate `AdapterLuxon` to our repository

### PR DESCRIPTION
**Pull request by flaviendelangle**

### CODE REVIEW

File: packages/x-date-pickers/src/AdapterDayjs/AdapterDayjs.test.tsx [hunk 1]
```diff
@@ -45,45 +45,45 @@ describe('<AdapterDayjs />', () => {
   describeGregorianAdapter(AdapterDayjs, { formatDateTime: 'YYYY-MM-DD HH:mm:ss' });
 
   describe('Adapter localization', () => {
```
Comment (flaviendelangle) [inline packages/x-date-pickers/src/AdapterDayjs/AdapterDayjs.test.tsx:47]: I just switched Ru and En tests to have the most common one at the top of the file

File: packages/x-date-pickers/src/AdapterDayjs/AdapterDayjs.ts [hunk 2]
```diff
@@ -231,28 +230,28 @@ export class AdapterDayjs implements MuiPickersAdapter<Dayjs> {
     return this.expandFormat(format).replace(/a/gi, '(a|p)m').toLocaleLowerCase();
   };
 
-  public isNull = (date: Dayjs | null) => {
```
Comment (flaviendelangle) [inline packages/x-date-pickers/src/AdapterDayjs/AdapterDayjs.ts:234]: Unified the param names on all the methods to match the JSDoc ones

File: packages/x-date-pickers/src/AdapterDayjs/AdapterDayjs.ts [hunk 3]
```diff
@@ -468,13 +467,13 @@ export class AdapterDayjs implements MuiPickersAdapter<Dayjs> {
...
   };
 
-  public getWeekNumber = (date: Dayjs) => {
-    return date.week();
+  public getWeekNumber = (value: Dayjs) => {
+    return value.week();
   };
 
   public getYearRange = (start: Dayjs, end: Dayjs) => {
-    const startDate = this.dayjs(start).startOf('year');
```
Comment (flaviendelangle) [inline packages/x-date-pickers/src/AdapterDayjs/AdapterDayjs.ts:476]: No need to re-parse here (same on `getWeekArray`), `startOf` and `endOf` are not mutating the initial date

File: packages/x-date-pickers/src/tests/describeGregorianAdapter/testLocalization.ts [hunk 4]
```diff
@@ -12,9 +12,11 @@ export const testLocalization: DescribeGregorianAdapterTestSuite = ({ adapter })
   });
 
   it('Method: getFormatHelperText', () => {
-    expect(adapter.getFormatHelperText(adapter.formats.keyboardDate)).to.equal('mm/dd/yyyy');
+    expect(adapter.getFormatHelperText(adapter.formats.keyboardDate)).to.equal(
+      adapter.lib === 'luxon' ? 'm/d/yyyy' : 'mm/dd/yyyy',
+    );
     expect(adapter.getFormatHelperText(adapter.formats.keyboardDateTime12h)).to.equal(
-      'mm/dd/yyyy hh:mm (a|p)m',
+      adapter.lib === 'luxon' ? 'm/d/yyyy hh:mm (a|p)m' : 'mm/dd/yyyy hh:mm (a|p)m',
```
Comment (flaviendelangle) [inline packages/x-date-pickers/src/tests/describeGregorianAdapter/testLocalization.ts:19]: This method was not working on Luxon on `date-io` so it's not a behavior change to change the test, it was just never used with Luxon :+1: 

By the way, the method is deprecated and we never use it

**Review by alexfauquette — APPROVED** [review]
Nice 👍

Kind of out of context: When doing the `data-fns` migration, it will be the opportunity to fix the following issues about typing of the `locale` attributes for adapters

https://github.com/mui/mui-x/issues/4598#issuecomment-1106382344

**Comment by flaviendelangle** [comment]
We should be able to type it correctly once every adapter is migrated indeed :ok_hand:

**Review by LukasTy — APPROVED** [review]
🙌 
Consider also removing `@date-io+luxon+2.16.1` as it's no longer relevant. 😉

**Comment by flaviendelangle** [comment]
@alexfauquette @LukasTy I'll wait 1-2 weeks to have feedback on the Dayjs adapter if the migration is breaking in a non-expected way :+1:

**Pull request merged**